### DISCUSSION TITLE
Network Error: A very annoying error

### DISCUSSION THREAD

**Discussion by hieutranduc** [urql-graphql/urql#1930] [Q&A] [answered]
If I open the app up and use it continuously, there is no problem with fetching data with `urql`. However, If I open the app and dont use it for about 30 minutes (or more), when I come back to use it, touch any element that needs to fetch data, then it takes a very long lasting fetch state **(~ 2 minutes)** and then reported Network Error. I added `retryExchange` but it doesn't work (I don't know how to check if it works or not). So I just can kill the app & reopen it to get everything back to normal.

I don't know where the problem is & how to fix it.

On server, I use serverless, so maybe after a period of time there is no request, server will be frozen, the next request will take a little more time to wake up, but it shouldnt take much time & before using ` urql`, I have no similar problem with `react-query`.

Can someone help me?

<img width="965" alt="Screen Shot 2021-09-06 at 11 14 36" src="https://user-images.githubusercontent.com/1744522/132230463-d23ca421-7a45-4bbe-a242-c5ca95dbbc6d.png">

![Simulator Screen Shot - iPhone 12 - 2021-09-06 at 11 12 23](https://user-images.githubusercontent.com/1744522/132230593-b0b55777-03a1-4a2d-ae05-b40dbd7f2ed1.png)

```ts
const client = createClient({
  url: 'https://api.run.app/graphql',
  exchanges: [
    devtoolsExchange,
    dedupExchange,
    customCacheExchange,

    authExchange({
      addAuthToOperation: ({ authState, operation }) => {
        if (!authState || !(authState as AuthState).token) {
          return operation;
        }

        // fetchOptions can be a function (See Client API) but you can simplify this based on usage
        const fetchOptions =
          typeof operation.context.fetchOptions === 'function'
            ? operation.context.fetchOptions()
            : operation.context.fetchOptions || {};

        return makeOperation(operation.kind, operation, {
          ...operation.context,
          fetchOptions: {
            ...fetchOptions,
            headers: {
              ...fetchOptions.headers,
              Authorization: `Bearer ${(authState as AuthState).token}`,
            },
          },
        });
      },
      getAuth: async ({ authState }) => {
        if (!authState) {
          const token = await auth().currentUser?.getIdToken();

          if (token) {
            return { token };
          }
        }

        return null;
      },
      didAuthError: ({ error }) => {
        console.log('didAuthError', error);
        return error.graphQLErrors.some((e) =>
          e.message.includes('permission denied'),
        );
      },
      willAuthError: ({ authState }) => {
        if (!authState) {
          return true;
        }
        // e.g. check for expiration, existence of auth etc
        return false;
      },
    }),
    retryExchange({
      initialDelayMs: 1000,
      maxDelayMs: 10000,
      randomDelay: false,
      maxNumberAttempts: 2,
      retryIf: (error) => {
        return !!(error.message.includes('Network') || error.networkError);
      },
    }),
    fetchExchange,
  ],
});
```

**Comment by hieutranduc** [comment]
@kitten can you help me?

**Comment by kitten** [comment, marked as answer]
I think this is very cut and dry simple. I can see in the screenshot that your API eventually reports back with a 503 and you'll probably want to look into that. In the retry exchange you've got a max attempt number that probably makes this plausibly fail just twice, but this ultimately comes down to your API and not any library you're using

**Reply by hieutranduc** [reply]
Thank you @kitten, your answer is absolutely USELESS, which helped me switch the whole project back to using `React Query` right away, without any further thought, because I knew there was nothing more I could do with `urql`. And everything is back to normal, smooth again. My GraphQL API Server is still working fine (for years, with Relay, with graphql-request, with React Query), no problems at all.

I can tell you the things that need to be improved right away with `urql`, but unfortunately I can't spend more time on `urql`.

Again, thank @kitten.

**Reply by kitten** [reply]
cool, mate, like I said, I can only look at the symptoms and tell you what I see, not work magic. so, thank you for your "USELESS" toxic feedback in a safe space where we're trying to help people for free and do it willingly because we'd like to push spaces, in this case, client-side GraphQL forward. I hope you reflect on this behaviour. I can understand frustration, especially facing issues that you aren't sure how to fix, yet I can't accept toxicity. You'll note that I tried to give you an answer the same day you asked the question, I ignored the ping comment, which was entirely unnecessary and simply impatient, then you didn't reply with further details for another 9 days.

This is a two way road, so please — with the utmost respect and love — reflect on this. https://github.com/FormidableLabs/urql/blob/main/CODE_OF_CONDUCT.md

I was willing to chat about this; in fact, if you would've responded with more comments on the above in that same hour, I would've likely loved to chat with you about chasing this issue down. However, since I don't want to expose you to more of my current and former colleagues on this repo and this was already too much for my taste, I'll have to give you a timeout/block from the repo for a day. Sorry!

**Reply by frederikhors** [reply]
If you don't know how to write an app you can pay someone to do it for you, but you don't have to offend people who have been working for years and day and night for a wonderful and gorgeous project like `urql`.

And above all, you can't name people like @kitten at all who always help when they can by answering with precision worthy of a thesis and fixing bugs in a matter of hours if not minutes.

Shame on you, @tranduchieu.
